src/gravel/controllers/kv.py

```diff
@@ -12,125 +12,395 @@
+                    # hangs before either callback is hit)
+                    # def _write_complete(comp):
+                    #     logger.debug(f"write complete {comp}")
+                    # def _write_safe(comp):
+                    #     logger.debug(f"write safe {comp}")
+                    # ioctx.operate_aio_write_op(op, "kvstore",
+                    #     _write_complete,
+                    #     _write_safe)
+                    # Q: can we determine upfront whether a write will block?
+                    self._ioctx.operate_write_op(op, "kvstore")
+                    # This next notifies all watchers *INCLUDING* me!
+                    self._ioctx.notify("kvstore", key)
+            except Exception as e:
+                # TODO: deal with this (log it? ignore it?)
+                # e.g. RADOS rados state (You cannot perform that operation on a Rados object in state configuring.)
+                logger.error(str(e))
+
+        logger.debug(f"Writing {key}: {value} to local cache")
+        self._db[key] = bvalue
+
+    def _get(self, key: str) -> Optional[str]:
+        # Try to get the value from the kvstore in our pool,
+        # if that works, stash it in our local cache, then
+        # return the value from the local cache.  If we can't
+        # get the value from the cluster, this gets whatever
+        # was last stashed in the local cache.
+        # This implies that it's possible for values to be
+        # quite stale in bizarre failure cases (value not read
+        # for a long time, then updated in cluster by some other
+        # instance, then cluster dies, then this instance reads,
+        # gets the old value)
+        if self._ioctx:
+            try:
+                with rados.ReadOpCtx() as op:
+                    omap_iter, ret = self._ioctx.get_omap_vals_by_keys(
+                        op, (key,)
+                    )
+                    assert ret == 0  # ???
+                    # TODO: does this need to be async?
+                    self._ioctx.operate_read_op(op, "kvstore")
+                    kv = dict(omap_iter)
+                    if key in kv:
+                        self._db[key] = kv[key]
+                    else:
+                        # key not present in cluster kvstore, make sure
+                        # it's also not present in db (prevent stale cache)
+                        if key in self._db:
+                            del self._db[key]
+
+            except Exception as e:
+                # TODO: deal with this (log it? ignore it?)
+                logger.error(str(e))
+
+        value = self._db.get(key)
         if not value:
             return None
         return value.decode("utf-8")
 
+    async def get(self, key: str) -> Optional[str]:
+        """Get value for provided key"""
+        # Wraps _get() so we can use _get() outside asyncio in _config_notify()
+        return self._get(key)
+
     async def get_prefix(self, key: str) -> List[str]:
```

Comment:
(nit the param here should probably be `prefix` not `key` (or perhaps `key_prefix`))

Reply:
I've changed it to key_prefix